Session.cc

```diff
@@ -229,10 +236,12 @@ void Session::onCloseFile(const CARTA::CloseFile& message, uint32_t requestId) {
     std::unique_lock<std::mutex> lock(frameMutex);
     if (fileId == ALL_FILES) {
         for (auto& frame : frames) {
+            frame.second->disconnect_called();
             frame.second.reset();  // delete Frame
```

Comment:
Does this give the Frame time to react to being disconnected (stop spectral profile) before deleting it?  I think it may still seg fault on Frame destruction.

Reply:
Same with main OnDisconnect -> Session::disconnect_called() -> Frame::disconnect_called(): Session is deleted after Session::disconnect_called(), and ~Session deletes Frame

Reply:
unless `disconnet_called` is doing some form of busy wait to ensure that the task has been concluded (either because it finished properly, or because it was interrupted correctly), then you'll still get a seg fault.

Reply:
It is true. Now I add a private variable `tbb::atomic<int> job_count_` to count the number of jobs in the Frame and use a while loop to wait for Frame's jobs finished before deleting it.

Reply:
Using the job count and wait for job count becomes zero before deleting an object looks work. However, the downside is that we have to carefully arrange the position of `increase_job_count_()` and `decrease_job_count_()` to make sure the job count goes to zero when all processes are finished. Also, the job count has to be counted immediately when the message comes in.

Reply:
I have not tested it but I think it looks logically correct. Normally I don't like busy waits but the way it is used here means that it will never wait long. The style is not consistent and job_count is not very descriptive. Also there is now a busy wait in the for frames in the Session destructor. This is OK, but since the Session destructor should never be called until all tasks processing on this Session are complete, it is confusing to have a busy wait here. If job count is not zero here it would indicate that a Session object is being destroyed due to a code bug.

Reply:
Thanks for the comments! I moved the wait for finish step into `Frame::DisconnectCalled()`. When `Session::disconnect_called()` is called, it will trigger `Frame::DisconnectCalled()` to execute stop and wait for finish step in Frames. So there is no busy wait in the Session destructor now. The code style I use is google's style, since Angus wants a unified style. But I can change them back to the consistent style before the decision is made. Also, I may change the name `job_counter` to `spectral_profile_counter`(?) I am wondering should I apply the "wait for finish step" on the other Frame's functions which are expected to have a long process time. Because they will also cause a crash like the spectral profile process. The coding is not finished yet and they need more checks. Any comments or suggestions are welcome.